clash-prelude/src/Clash/Class/Resize.hs

```diff
@@ -37,3 +47,65 @@ class Resize (f :: Nat -> Type) where
   signExtend = resize
   -- | Remove bits from the MSB
   truncateB :: KnownNat a => f (a + b) -> f a
+
+-- | Helper function of 'fromIntegralChecked' and 'resizeChecked'
+checkIntegral ::
+  forall a b.
+  HasCallStack =>
+  (Integral a, Integral b, Bounded b) =>
+  Proxy b ->
+  a -> ()
+checkIntegral Proxy a =
+  if toInteger a > toInteger (maxBound @b)
+  || toInteger a < toInteger (minBound @b) then
+    error $ "Given integral " <> show (toInteger a) <> " is out of bounds for" <>
+            " target type. Bounds of target type are: [" <>
+            show (toInteger (minBound @b)) <> ".." <>
+            show (toInteger (maxBound @b)) <> "]."
+  else
+    ()
+
+-- | Like 'fromIntegral', but errors if /a/ is out of bounds for /b/. Useful when
+-- you "know" /a/ can't be out of bounds, but would like to have your assumptions
+-- checked.
+--
+-- __N.B.__: Check only affects simulation. I.e., no checks will be inserted
+-- into the generated HDL
+checkedFromIntegral::
+  forall a b.
+  HasCallStack =>
+  (Integral a, Integral b, Bounded b) =>
+  a -> b
+checkedFromIntegral a =
+  checkIntegral (Proxy @b) a `seq` fromIntegral a
+
+-- | Like 'resize', but errors if /f a/ is out of bounds for /f b/. Useful when
+-- you "know" /f a/ can't be out of bounds, but would like to have your
+-- assumptions checked.
+--
+-- __N.B.__: Check only affects simulation. I.e., no checks will be inserted
+-- into the generated HDL
+checkedResize ::
+  forall a b f.
+  ( HasCallStack
+  , Resize f
+  , KnownNat a, Integral (f a)
+  , KnownNat b, Integral (f b), Bounded (f b) ) =>
+  f a -> f b
+checkedResize a =
+  checkIntegral (Proxy @(f b)) a `seq` resize a
+
+-- | Like 'truncateB', but errors if /f (a + b)/ is out of bounds for /f a/. Useful
+-- when you "know" /f (a + b)/ can't be out of bounds, but would like to have your
+-- assumptions checked.
+--
+-- __N.B.__: Check only affects simulation. I.e., no checks will be inserted
+-- into the generated HDL
+checkedTruncateB ::
+  forall a b f.
+  ( HasCallStack
+  , Resize f
+  , KnownNat b, Integral (f (a + b))
+  , KnownNat a, Integral (f a), Bounded (f a) ) =>
+  f (a + b) -> f a
+checkedTruncateB = checkedResize
```

Comment:
This calls `resize` indirectly, shouldn't it be calling `truncateB`?

Reply:
It _should_ be the same thing. IIRC `resize` calls either `extend` or `truncateB` depending on its type arguments.

Reply:
I'll double check

Reply:
I changed it to `truncateB` so I don't have to double check :-)